internal/k8s/reconciler/state/gateway.go

```diff
@@ -0,0 +1,104 @@
+package state
+
+import (
+	"errors"
+
+	"github.com/hashicorp/consul-api-gateway/internal/core"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/reconciler/common"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/reconciler/status"
+	gw "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// GatewayState holds ephemeral state for gateways
+type GatewayState struct {
+	ConsulNamespace string
+	Status          status.GatewayStatus
+	PodReady        bool
+	ServiceReady    bool
+	Generation      int64
+	Addresses       []string
+
+	Listeners []*ListenerState
+}
+
+func InitialGatewayState(g *gw.Gateway) *GatewayState {
+	state := &GatewayState{
+		Generation: g.GetGeneration(),
+	}
+	for _, listener := range g.Spec.Listeners {
+		state.Listeners = append(state.Listeners, &ListenerState{
+			Name:     listener.Name,
+			Protocol: listener.Protocol,
+			Routes:   make(map[string]core.ResolvedRoute),
+		})
+	}
+	return state
+}
+
+func (g *GatewayState) GetStatus(gateway *gw.Gateway) gw.GatewayStatus {
+	listenerStatuses := []gw.ListenerStatus{}
+	for _, state := range g.Listeners {
+		listenerStatuses = append(listenerStatuses, state.getStatus(g.Generation))
+	}
+
+	conditions := g.Status.Conditions(g.Generation)
+
+	// prefer to not update to not mess up timestamps
+	if status.ListenerStatusesEqual(listenerStatuses, gateway.Status.Listeners) {
+		listenerStatuses = gateway.Status.Listeners
+	}
+	if status.ConditionsEqual(conditions, gateway.Status.Conditions) {
+		conditions = gateway.Status.Conditions
+	}
+
+	ipType := gw.IPAddressType
+	addresses := make([]gw.GatewayAddress, 0, len(g.Addresses))
+	for _, address := range g.Addresses {
+		addresses = append(addresses, gw.GatewayAddress{
```

Comment:
nit: if you're going to create a slice of length `len(g.Addresses)` might as well skip the append and just index off the for loop